Bot webhook (#34)
callbellchat/callbell-public-docs: i18n/pt/docusaurus-plugin-content-docs/current/bot/editor/actions/logic/webhook.md

    <th>Ação</th>
  </tr>
  <tr>
    <td>MissingParameter</td>
    <td>Url ou método ausente</td>
    <td>Verifique novamente se o URL e o método estão especificados corretamente.</td>
  </tr>
  <tr>
    <td>BadRequest</td>
    <td>URL, corpo, parâmetros ou cabeçalho inválidos</td>
    <td>Verifique a validade de cada parâmetro. Use clientes como o Postman para verificar seus parâmetros.</td>
  </tr>
  <tr>
    <td>BadResponse</td>
    <td>Status da resposta diferente de 200 ou a resposta não está no formato JSON</td>
    <td>Verifique o status de seu servidor.</td>
  </tr>
  <tr>
    <td>KeyError</td>
    <td>O caminho especificado não corresponde à carga útil da resposta</td>
    <td>Verifique novamente a validade do caminho que você inseriu na ação do webhook para cada variável</td>
  </tr>
</table>

### Analisar e armazenar a resposta das solicitações em variáveis

Quando sua solicitação receber uma resposta bem-sucedida, você poderá escolher uma variável que armazenará o resultado. Você também pode selecionar a parte da resposta que deseja armazenar usando o auxiliar "path" (caminho):

<div class="text--center">
    <img src={pathHelper} width={500} />
</div>

> Se estiver tentando selecionar uma matriz como valor, o helper sugerirá o uso da função `flatMap`, que lhe dá a possibilidade de extrair qualquer valor da matriz.

Reviewer: ```suggestion
Se estiver tentando selecionar um arranjo (_array_) como valor, o helper sugerirá o uso da função `flatMap`, que lhe dá a possibilidade de extrair qualquer valor dele.
```

Por exemplo, digamos que a estrutura de dados da resposta de sua solicitação seja semelhante a esta:

```
{
  "products": [
    {
      "id": 1,
      "title": "iPhone 9",
      "description": "Um celular da apple que não é nada parecido com a apple",
      "price": 549,
      "rating": 4.69,
      "estoque": 94,
      "marca": "Apple",
      "category": "smartphones",
      "images": [
        "https://cdn.dummyjson.com/product-images/1/1.jpg",
        "https://cdn.dummyjson.com/product-images/1/2.jpg"
      ]
    },
    {
      "id": 2,
      "title": "iPhone X",
      "description": "Sem SIM, modelo A19211 Tela Super Retina HD de 6,5 polegadas com tecnologia OLED Chip A12 Bionic com ...",
      "price": 899,
      "rating": 4.44,
      "estoque": 34,
      "marca": "Apple",
      "category": "smartphones",
      "images": [
        "https://cdn.dummyjson.com/product-images/2/1.jpg"
      ]
    }
  ]
}
```

> E você deseja armazenar na variável `webhook success` uma matriz somente de títulos de produtos:

Reviewer: ```suggestion
E você deseja armazenar na variável `webhook success` um _array_ somente de títulos de produtos:
```

```
["iPhone 9", "iPhone X"]
```

O uso do valor a seguir atenderá às suas necessidades:

<div class="text--center">
    <img src={flatMapExemple} width={500} />
</div>

Se quiser acessar a categoria do segundo item do produto, você pode usar esse caminho:

```
products[1].category
```

:::tip
Você pode armazenar várias partes da resposta em diferentes variáveis. Isso evitará que você execute a mesma solicitação de webhook várias vezes

<div class="text--center">
    <img src={multipleResponseVariables} width={500} />
</div>
:::

### Lidando com solicitações com falha

É possível que, mesmo depois de configurar corretamente sua ação de webhook, ela falhe ocasionalmente (erros de servidor, parâmetros não tratados corretamente etc.).

> Se o comportamento do seu bot depender do resultado da solicitação, é importante lidar com os erros.

Reviewer: ```suggestion
Se o comportamento do seu bot depender do resultado da solicitação, é importante dar uma tratativa aos erros.
```

O bot armazenará todos os erros na variável que você especificou ou usará a variável `last webhook failure`:

<div class="text--center">
    <img src={responseFailureVariable} width={500} />
</div>

Cabe a você tratar esse erro ou não. A melhor maneira é criar uma variável específica para cada webhook e, em seguida, verificar por meio de um bloco de opções se a variável contém uma mensagem:

<div class="text--center">
    <img src={failureHandling} width={600} />
</div>

## Teste sua solicitação no simulador

Você deve sempre verificar se o seu bot está se comportando conforme o esperado por meio do simulador, garantindo que as solicitações bem-sucedidas e com falha sejam tratadas corretamente.

> Quando o simulador chegar a uma ação de webhook, ao contrário do bot real, ele fará uma pausa e solicitará que você especifique a aparência da solicitação:

Reviewer: ```suggestion
Quando o simulador chegar a uma ação de webhook, ao contrário do bot real, ele fará uma pausa e solicitará que você especifique como deveria ser a solicitação:
```

<div class="text--center">
    <img src={simulatorWebhookChoices} width={500} />
</div>

> - Sucesso falso: Ele solicitará um modal para especificar a carga útil da solicitação bem-sucedida. Em seguida, ele utilizará cada caminho que você especificou na ação do webhook para analisar e armazenar as variáveis correspondentes. Observe que, se você estiver armazenando em uma variável do tipo lista, o simulador esperará que a carga útil fornecida contenha uma matriz no caminho correto. Caso contrário, será exibido um erro no modal, impedindo a progressão.

Reviewer: ```suggestion
- Sucesso falso: Ele solicitará um modal para especificar o _payload_ da solicitação bem-sucedida. Em seguida, ele utilizará cada caminho que você especificou na ação do webhook para analisar e armazenar as variáveis correspondentes. Observe que, se você estiver armazenando em uma variável do tipo lista, o simulador esperará que o _payload_ fornecido contenha um arranjo no caminho correto. Caso contrário, será exibido um erro no modal, impedindo a continuidade.
```

<div class="text--center">
    <img src={fakeSuccessWebhook} width={500} />
</div>

> - Falha falsa: Exibirá o mesmo modal e permitirá que você especifique a carga útil a ser salva na variável de falha especificada na ação.

Reviewer: ```suggestion
- Falha falsa: Exibirá o mesmo modal e permitirá que você especifique a carga útil a ser salva na variável de falha especificada na ação.
```
```suggestion
- Falha falsa: Exibirá o mesmo modal e permitirá que você especifique o _payload_ a ser salvo na variável de falha especificada na ação.
```
